Add page to create channel

rhysyngsun opened the pull request:
#### What are the relevant tickets?
Fixes #16 

#### What's this PR do?
Adds a page to create a channel

#### How should this be manually tested?
Go to http://localhost:8063/manage/channel/new and fill in the form. Verify you're redirected to your new channel.

#### Any background context you want to provide?
No validation yet

aliceriot commented on static/js/containers/admin/AdminPage.js:
```diff
@@ -0,0 +1,26 @@
+    return (
+      <div className="admin double-column">
+        <Card>
+          <h1>Admin</h1>
```
should we maybe add a `title` prop to the `Card` component? I'm thinking we may want to style that in a consistent way across the app (just occurred to me, not necessary something to do in this PR)

rhysyngsun replied:
Yeah, I think something like that would be good eventually.

noisecapella commented:
After I successfully create a channel I'm redirected to the channel page but I get an error page:

![screenshot from 2017-07-31 14-31-08](https://user-images.githubusercontent.com/863262/28792256-047ff94e-75fd-11e7-836f-a631c9d21b49.png)

noisecapella commented on static/js/components/admin/ChannelEditForm.js:
```diff
@@ -0,0 +1,76 @@
+  props: {
+    onSubmit: Function,
+    onUpdate: Function,
+    channel:  ?Object,
```
Can you use the `Channel` type here?

noisecapella commented on static/js/components/admin/ChannelEditForm_test.js:
```diff
@@ -0,0 +1,88 @@
+
+    describe('onUpdate', () => {
+      for (let inputName of ['title', 'name', 'public_description']) {
+        it(`should be called when ${inputName} input is modifed`, () => {
```
Should be `modified`

noisecapella commented on static/js/containers/admin/AdminPage.js:
```diff
@@ -0,0 +1,26 @@
+      <div className="admin double-column">
+        <Card>
+          <h1>Admin</h1>
+          <Route path={`${match.url}/channel/new`} component={CreateChannelPage}/>
```
Should we use the functions in `lib/url.js` instead for this kind of thing?

rhysyngsun replied:
Not in this case because it's the route config, not a link to a page.

noisecapella commented on static/js/containers/admin/CreateChannelPage.js:
```diff
@@ -0,0 +1,61 @@
+    e.preventDefault();
+
+    dispatch(actions.channels.post(channelForm.value)).then((channel) => {
+      history.push(`/channel/${channel.name}`);
```
Can we make a function in `lib/url.js` to create this URL?

noisecapella commented on static/js/lib/api.js:
```diff
@@ -17,6 +18,15 @@ export function getChannel(channelName: string): Promise<Channel> {
+  return fetchJSONWithCSRF(`/api/v0/channels/`, {
+    method: POST,
+    body: JSON.stringify({
+      ...channel
```
Why are we using the spread syntax here?

rhysyngsun replied:
Yeah, that's less than useful, it also isn't guarding against the other channel properties which could cause a 4xx.

noisecapella commented on static/js/lib/channels_test.js:
```diff
@@ -0,0 +1,15 @@
+      'title': '',
+      'name': '',
+      'public_description': '',
+      'channel_type': 'public',
```
Can we make constants for `public` and `private`?

noisecapella commented on static/js/lib/form_actions.js:
```diff
@@ -0,0 +1,46 @@
+    case actionTypes.reset:
+    case actionTypes.create:
+      return {
+        ...state,
```
What about `errors`? It seems like we should clear it here

rhysyngsun replied:
`errors` is covered by the spread, should I get rid of the spread and just be explicit about it?

noisecapella replied:
It's not the explicitness, I'm wondering about if we should replace `errors` here with `{}` to clear the errors

rhysyngsun replied:
Good point.

rhysyngsun replied:
I won't have a good way of testing that errors gets reset until I add support for that, which will be in a separate PR, but worth fixing this issue now.

noisecapella commented on static/js/lib/form_actions_test.js:
```diff
@@ -0,0 +1,103 @@
+      sandbox.restore();
+    });
+
+    //
```
Can you remove this line?

rhysyngsun commented:
@noisecapella the diff is probably hosed now since I rebased. Look at this commit to see what I fixed relative your feedback: https://github.com/mitodl/open-discussions/pull/73/commits/ce770eccb011bf87b3e1785b24cef3d08e837155

rhysyngsun commented:
@noisecapella didn't see the error until just now. This would've been caused by an error from the backend so you might have to check your network tab or backend logs for more details

noisecapella commented:
I just checked out the newest commit, it looks like the error is gone now
